File: kotlin/HyperCycleClient.kt
// HyperCycleClient.kt — HyperCycle AIM Client SDK (Kotlin/Android)
// Version: 0.2.0-beta
//
// Generic client for connecting to any public AIM on any HyperCycle node.
// Designed for Android app developers building on HyperCycle.
//
// Dependencies (add to build.gradle):
//   implementation("com.squareup.okhttp3:okhttp:4.12.0")
//   implementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:1.7.3")
//
// All methods are suspend functions — call from a coroutine or use
// lifecycleScope / viewModelScope. Never call from the main thread directly.
//
// The node URL is read from the HYPERCYCLE_NODE_URL system property,
// or passed directly. Do not hardcode node IPs in shipping apps.
//
// Quick start:
//   val client = HyperCycleClient(nodeUrl = BuildConfig.NODE_URL)
//   val aim    = client.discover("yamnet-classifier")
//   if (aim is HyperCycleResult.Success) {
//       val result = client.execute(aim.data.slot, "infer", body)
//   }

package ai.hypercycle.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

const val SDK_VERSION = "0.2.0-beta"

// ---------------------------------------------------------------------------
// Result type
// ---------------------------------------------------------------------------

sealed class HyperCycleResult<out T> {
    data class Success<T>(val data: T, val statusCode: Int = 200) : HyperCycleResult<T>()
    data class Failure(val error: String, val statusCode: Int? = null) : HyperCycleResult<Nothing>()
}

// ---------------------------------------------------------------------------
// Endpoint cost info
// ---------------------------------------------------------------------------

data class EndpointCost(
    val currency:      String,
    val fixed:         Double? = null,
    val estimatedCost: Double? = null,
    val min:           Double? = null,
    val max:           Double? = null,
)

// ---------------------------------------------------------------------------
// AIM info — from /info endpoint
// ---------------------------------------------------------------------------

/**
 * Metadata for a deployed AIM, sourced from the node's /info endpoint.
 *
 * The /info endpoint is the primary discovery mechanism for running AIMs.
 * [slot] and [imageName] are the key fields for app integration.
 */
data class AIMInfo(
    val slot:        Int,
    val port:        Int,
    val imageName:   String,
    val imageTag:    String,
    val status:      String,   // "running" = callable
    val labels:      Map<String, String>,
    val containerID: String,
    val costs:       List<EndpointCost>,
)

// ---------------------------------------------------------------------------
// Node info — from /info endpoint
// ---------------------------------------------------------------------------

/**
 * Full node info from the /info endpoint.
 *
 * This is the primary entry point for any app integrating with a HyperCycle
 * node. The [aims] list contains every deployed AIM with its metadata.
 */
data class NodeInfo(
    val status:              String,
    val name:                String,
    val address:             String,
    val nodeId:              String,
    val network:             String,
    val nodeVersion:         String,
    val platform:            String,
    val aims:                List<AIMInfo>,
    val acceptingCurrencies: List<String>,
    val raw:                 JSONObject,   // full /info response
)

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

class HyperCycleClient(
    nodeUrl: String? = null,
    timeoutSeconds: Long = 30L,
) {
    companion object {
        const val SDK_VERSION = ai.hypercycle.sdk.SDK_VERSION
    }

    private val JSON_TYPE = "application/json; charset=utf-8".toMediaType()

    private val baseUrl: String = run {
        val url = nodeUrl
            ?: System.getProperty("HYPERCYCLE_NODE_URL")
            ?: System.getenv("HYPERCYCLE_NODE_URL")
        requireNotNull(url) {
            "nodeUrl is required. Pass it directly or set HYPERCYCLE_NODE_URL."
        }
        url.trimEnd('/')
    }

    private val http = OkHttpClient.Builder()
        .connectTimeout(timeoutSeconds, TimeUnit.SECONDS)
        .readTimeout(timeoutSeconds, TimeUnit.SECONDS)
        .writeTimeout(timeoutSeconds, TimeUnit.SECONDS)
        .build()

    // -------------------------------------------------------------------------
    // Node info
    // -------------------------------------------------------------------------

    /**
     * GET /info — fetch full node info including all deployed AIMs.
     *
     * This is the primary discovery endpoint. Call this (or [discover]) before
     * calling [health] or [execute].
     *
     * Must be called from a coroutine (suspend function).
     */
    suspend fun info(): HyperCycleResult<NodeInfo> = withContext(Dispatchers.IO) {
        val raw = get("/info")
        if (raw is HyperCycleResult.Failure) return@withContext raw

        val data = (raw as HyperCycleResult.Success).data
        val aimBlock = data.optJSONObject("aim")
        val aimsJson = aimBlock?.optJSONArray("aims")
            ?: return@withContext HyperCycleResult.Failure("Invalid /info response: missing aim.aims")

        val aims = (0 until aimsJson.length()).map { parseAIM(aimsJson.getJSONObject(it)) }
        val currencies = data.optJSONArray("accepting_currencies")?.let { arr ->
            (0 until arr.length()).map { arr.getString(it) }
        } ?: emptyList()

        HyperCycleResult.Success(NodeInfo(
            status              = data.optString("status", ""),
            name                = data.optString("name", ""),
            address             = data.optString("address", ""),
            nodeId              = data.optString("node_id", ""),
            network             = data.optString("network", ""),
            nodeVersion         = data.optString("node_version", ""),
            platform            = data.optString("platform", ""),
            aims                = aims,
            acceptingCurrencies = currencies,
            raw                 = data,
        ))
    }

    // -------------------------------------------------------------------------
    // Discovery
    // -------------------------------------------------------------------------

    /**
     * Find a deployed, running AIM by Docker image name.
     *
     * @param imageName Docker image name, e.g. "yamnet-classifier"
     */
    suspend fun discover(imageName: String): HyperCycleResult<AIMInfo> {
        val infoResult = info()
        if (infoResult is HyperCycleResult.Failure) {
            return HyperCycleResult.Failure(
                "Failed to fetch node info: ${infoResult.error}", infoResult.statusCode
            )
        }
        val aims = (infoResult as HyperCycleResult.Success).data.aims
        val match = aims.find { it.imageName == imageName }
            ?: run {
                val available = aims.map { it.imageName }
                return HyperCycleResult.Failure(
                    "AIM '$imageName' not found on node. Available: $available"
                )
            }

        if (match.status != "running") {
            return HyperCycleResult.Failure(
                "AIM '$imageName' found at slot ${match.slot} but status is '${match.status}', not 'running'."
            )
        }
        return HyperCycleResult.Success(match)
    }

    /**
     * Return metadata for all AIMs on the node.
     *
     * @param runningOnly If true (default), return only AIMs with status "running".
     */
    suspend fun discoverAll(runningOnly: Boolean = true): HyperCycleResult<List<AIMInfo>> {
        val infoResult = info()
        if (infoResult is HyperCycleResult.Failure) {
            return HyperCycleResult.Failure(infoResult.error, infoResult.statusCode)
        }
        val aims = (infoResult as HyperCycleResult.Success).data.aims
        return HyperCycleResult.Success(
            if (runningOnly) aims.filter { it.status == "running" } else aims
        )
    }

    // -------------------------------------------------------------------------
    // Health
    // -------------------------------------------------------------------------

    /**
     * GET /aim/<slot>/health — check AIM liveness and model readiness.
     *
     * @param slot AIM slot number (from [discover] or [info])
     */
    suspend fun health(slot: Int): HyperCycleResult<JSONObject> =
        withContext(Dispatchers.IO) { get("/aim/$slot/health") }

    // -------------------------------------------------------------------------
    // Cost estimation
    // -------------------------------------------------------------------------

    /**
     * POST with cost_only header — returns estimated cost without executing.
     * No charge. No signature required.
     *
     * When an AIM has zero-cost endpoints (e.g. open beta), all values are 0.
     *
     * @param slot      AIM slot number
     * @param endpoint  Endpoint name, e.g. "infer"
     * @param body      Full request body (same shape as [execute])
     */
    suspend fun estimate(
        slot:     Int,
        endpoint: String,
        body:     JSONObject,
    ): HyperCycleResult<JSONObject> = withContext(Dispatchers.IO) {
        post(
            path         = "/aim/$slot/$endpoint",
            body         = body,
            extraHeaders = mapOf("cost_only" to "true"),
        )
    }

    // -------------------------------------------------------------------------
    // Execution
    // -------------------------------------------------------------------------

    /**
     * POST /aim/<slot>/<endpoint> — execute an AIM endpoint.
     *
     * The Node Manager authenticates, meters, and routes this to the AIM.
     * For zero-cost AIMs no wallet signature or balance is required.
     *
     * @param slot      AIM slot number
     * @param endpoint  Endpoint name, e.g. "infer"
     * @param body      Request body — will be JSON serialized
     */
    suspend fun execute(
        slot:     Int,
        endpoint: String,
        body:     JSONObject,
    ): HyperCycleResult<JSONObject> = withContext(Dispatchers.IO) {
        post("/aim/$slot/$endpoint", body)
    }

    // -------------------------------------------------------------------------
    // Convenience: ping
    // -------------------------------------------------------------------------

    /** Quick connectivity check. Returns true if the node is reachable. */
    suspend fun ping(): Boolean {
        val result = withContext(Dispatchers.IO) { get("/info") }
        return result is HyperCycleResult.Success
    }

    // -------------------------------------------------------------------------
    // Internal helpers
    // -------------------------------------------------------------------------

    private fun parseAIM(a: JSONObject): AIMInfo {
        val costs = mutableListOf<EndpointCost>()
        a.optJSONObject("uri_cost")?.let { uriCost ->
            for (endpointKey in uriCost.keys()) {
                uriCost.optJSONObject(endpointKey)?.let { currencies ->
                    for (currency in currencies.keys()) {
                        currencies.optJSONObject(currency)?.let { cd ->
                            costs.add(EndpointCost(
                                currency      = cd.optString("currency", currency),
                                fixed         = if (cd.has("fixed"))         cd.getDouble("fixed")         else null,
                                estimatedCost = if (cd.has("estimated_cost")) cd.getDouble("estimated_cost") else null,
                                min           = if (cd.has("min"))           cd.getDouble("min")           else null,
                                max           = if (cd.has("max"))           cd.getDouble("max")           else null,
                            ))
                        }
                    }
                }
            }
        }
        val labelsObj = a.optJSONObject("labels")
        val labels = labelsObj?.keys()?.asSequence()
            ?.associateWith { labelsObj.optString(it) } ?: emptyMap()

        return AIMInfo(
            slot        = a.optInt("slot", 0),
            port        = a.optInt("port", 0),
            imageName   = a.optString("image_name", ""),
            imageTag    = a.optString("image_tag", "latest"),
            status      = a.optString("status", "unknown"),
            labels      = labels,
            containerID = a.optString("container_id", ""),
            costs       = costs,
        )
    }

    private fun get(path: String): HyperCycleResult<JSONObject> {
        val request = Request.Builder().url("$baseUrl$path").get().build()
        return executeRequest(request)
    }

    private fun post(
        path:         String,
        body:         JSONObject,
        extraHeaders: Map<String, String> = emptyMap(),
    ): HyperCycleResult<JSONObject> {
        val requestBody = body.toString().toRequestBody(JSON_TYPE)
        val builder     = Request.Builder().url("$baseUrl$path").post(requestBody)
        extraHeaders.forEach { (k, v) -> builder.addHeader(k, v) }
        return executeRequest(builder.build())
    }

    private fun executeRequest(request: Request): HyperCycleResult<JSONObject> {
        return try {
            val response   = http.newCall(request).execute()
            val bodyString = response.body?.string()
                ?: return HyperCycleResult.Failure("Empty response", response.code)
            val json = try {
                JSONObject(bodyString)
            } catch (e: Exception) {
                return HyperCycleResult.Failure(
                    "Invalid JSON response: $bodyString", response.code
                )
            }
            if (!response.isSuccessful) {
                return HyperCycleResult.Failure(
                    "HTTP ${response.code} from ${request.url}: $bodyString",
                    response.code
                )
            }
            HyperCycleResult.Success(json, response.code)
        } catch (e: IOException) {
            HyperCycleResult.Failure("Request failed: ${e.message}")
        }
    }
}

File: swift/HyperCycleClient.swift
// HyperCycleClient.swift — HyperCycle AIM Client SDK (Swift/iOS)
// Version: 0.3.1-beta
//
// Generic client for connecting to any public AIM on any HyperCycle node.
// Designed for iOS (and macOS) app developers building on HyperCycle.
//
// Requires iOS 15+ / macOS 12+ (async/await + URLSession).
// Zero dependencies — Foundation only.
//
// The node URL is read from the HYPERCYCLE_NODE_URL environment variable,
// or passed directly. Do not hardcode node IPs in shipping apps.
//
// Quick start:
//   let client = HyperCycleClient()   // reads HYPERCYCLE_NODE_URL
//   let aim    = try await client.discover(imageName: "yamnet-classifier").get()
//   let health = try await client.health(slot: aim.slot).get()
//   let result = try await client.execute(slot: aim.slot, endpoint: "infer", body: body).get()

import Foundation

// ---------------------------------------------------------------------------
// Version
// ---------------------------------------------------------------------------

public let SDK_VERSION = "0.3.1-beta"

// ---------------------------------------------------------------------------
// Result type
// ---------------------------------------------------------------------------

public enum HyperCycleResult<T> {
    case success(T)
    case failure(HyperCycleError)
}

public struct HyperCycleError: Error {
    public let message: String
    public let statusCode: Int?

    init(_ message: String, statusCode: Int? = nil) {
        self.message    = message
        self.statusCode = statusCode
    }
}

// ---------------------------------------------------------------------------
// Endpoint cost info
// ---------------------------------------------------------------------------

public struct EndpointCost {
    public let currency:      String
    public let fixed:         Double?
    public let estimatedCost: Double?
    public let min:           Double?
    public let max:           Double?
}

// ---------------------------------------------------------------------------
// AIM info — from /info endpoint
// ---------------------------------------------------------------------------

/// Metadata for a deployed AIM, sourced from the node's /info endpoint.
///
/// The /info endpoint is the primary discovery mechanism for running AIMs.
/// `slot` and `imageName` are the key fields for app integration.
public struct AIMInfo {
    public let slot:        Int
    public let port:        Int
    public let imageName:   String
    public let imageTag:    String
    public let status:      String      // "running" = callable
    public let labels:      [String: String]
    public let containerID: String
    public let costs:       [EndpointCost]
}

// ---------------------------------------------------------------------------
// Node info — from /info endpoint
// ---------------------------------------------------------------------------

/// Full node info from the /info endpoint.
///
/// This is the primary entry point for any app integrating with a HyperCycle
/// node. The `aims` array contains every deployed AIM with its metadata.
public struct NodeInfo {
    public let status:               String
    public let name:                 String
    public let address:              String
    public let nodeID:               String
    public let network:              String
    public let nodeVersion:          String
    public let platform:             String
    public let aims:                 [AIMInfo]
    public let acceptingCurrencies:  [String]
    public let raw:                  [String: Any]   // full /info response
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

public class HyperCycleClient {

    public static let sdkVersion = SDK_VERSION

    private let nodeURL:         String
    private let timeoutInterval: TimeInterval
    private let session:         URLSession

    /// - Parameters:
    ///   - nodeURL: Base URL of the Node Manager, e.g. "http://192.168.1.10:8000"
    ///              If nil, reads HYPERCYCLE_NODE_URL from the environment.
    ///   - timeout: Request timeout in seconds (default 30)
    public init(nodeURL: String? = nil, timeout: TimeInterval = 30) throws {
        let url = nodeURL ?? ProcessInfo.processInfo.environment["HYPERCYCLE_NODE_URL"]
        guard let url = url, !url.isEmpty else {
            throw HyperCycleError(
                "nodeURL is required. Pass it directly or set HYPERCYCLE_NODE_URL."
            )
        }
        self.nodeURL         = url.hasSuffix("/") ? String(url.dropLast()) : url
        self.timeoutInterval = timeout
        let config           = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        self.session         = URLSession(configuration: config)
    }

    // -------------------------------------------------------------------------
    // Node info
    // -------------------------------------------------------------------------

    /// GET /info — fetch full node info including all deployed AIMs.
    ///
    /// This is the primary discovery endpoint. Call this (or discover()) before
    /// calling health() or execute().
    public func info() async -> HyperCycleResult<NodeInfo> {
        let raw = await get(path: "/info")
        guard case .success(let data) = raw else {
            if case .failure(let err) = raw { return .failure(err) }
            return .failure(HyperCycleError("Failed to fetch /info"))
        }

        guard
            let aimBlock = data["aim"] as? [String: Any],
            let aimsRaw  = aimBlock["aims"] as? [[String: Any]]
        else {
            return .failure(HyperCycleError("Invalid /info response: missing aim.aims"))
        }

        let aims = aimsRaw.map { self.parseAIM($0) }

        return .success(NodeInfo(
            status:              data["status"]               as? String ?? "",
            name:                data["name"]                 as? String ?? "",
            address:             data["address"]              as? String ?? "",
            nodeID:              data["node_id"]              as? String ?? "",
            network:             data["network"]              as? String ?? "",
            nodeVersion:         data["node_version"]         as? String ?? "",
            platform:            data["platform"]             as? String ?? "",
            aims:                aims,
            acceptingCurrencies: data["accepting_currencies"] as? [String] ?? [],
            raw:                 data
        ))
    }

    // -------------------------------------------------------------------------
    // Discovery
    // -------------------------------------------------------------------------

    /// Find a deployed, running AIM by Docker image name.
    /// - Parameter imageName: Docker image name, e.g. "yamnet-classifier"
    public func discover(imageName: String) async -> HyperCycleResult<AIMInfo> {
        let infoResult = await info()
        guard case .success(let nodeInfo) = infoResult else {
            if case .failure(let err) = infoResult {
                return .failure(HyperCycleError(
                    "Failed to fetch node info: \(err.message)", statusCode: err.statusCode
                ))
            }
            return .failure(HyperCycleError("Failed to fetch node info"))
        }

        guard let match = nodeInfo.aims.first(where: { $0.imageName == imageName }) else {
            let available = nodeInfo.aims.map { $0.imageName }
            return .failure(HyperCycleError(
                "AIM '\(imageName)' not found on node. Available: \(available)"
            ))
        }

        guard match.status == "running" else {
            return .failure(HyperCycleError(
                "AIM '\(imageName)' found at slot \(match.slot) but status is '\(match.status)', not 'running'."
            ))
        }

        return .success(match)
    }

    /// Return metadata for all AIMs on the node.
    /// - Parameter runningOnly: If true (default), return only AIMs with status "running".
    public func discoverAll(runningOnly: Bool = true) async -> HyperCycleResult<[AIMInfo]> {
        let infoResult = await info()
        guard case .success(let nodeInfo) = infoResult else {
            if case .failure(let err) = infoResult { return .failure(err) }
            return .failure(HyperCycleError("Failed to fetch node info"))
        }
        let aims = runningOnly
            ? nodeInfo.aims.filter { $0.status == "running" }
            : nodeInfo.aims
        return .success(aims)
    }

    // -------------------------------------------------------------------------
    // Health
    // -------------------------------------------------------------------------

    /// GET /aim/<slot>/health — check AIM liveness and model readiness.
    /// - Parameter slot: AIM slot number (from discover() or info())
    public func health(slot: Int) async -> HyperCycleResult<[String: Any]> {
        return await get(path: "/aim/\(slot)/health")
    }

    // -------------------------------------------------------------------------
    // Cost estimation
    // -------------------------------------------------------------------------

    /// POST with cost_only header — returns estimated cost without executing.
    /// No charge. No signature required.
    ///
    /// When an AIM has zero-cost endpoints (e.g. open beta), all cost values
    /// will be 0 across all currencies.
    ///
    /// - Parameters:
    ///   - slot:     AIM slot number
    ///   - endpoint: Endpoint name, e.g. "infer"
    ///   - body:     Full request body (same shape as execute)
    public func estimate(
        slot:     Int,
        endpoint: String,
        body:     [String: Any]
    ) async -> HyperCycleResult<[String: Any]> {
        return await post(
            path:         "/aim/\(slot)/\(endpoint)",
            body:         body,
            extraHeaders: ["cost_only": "true"]
        )
    }

    // -------------------------------------------------------------------------
    // Execution
    // -------------------------------------------------------------------------

    /// POST /aim/<slot>/<endpoint> — execute an AIM endpoint.
    ///
    /// The Node Manager authenticates, meters, and routes this to the AIM.
    /// For zero-cost AIMs no wallet signature or balance is required.
    ///
    /// - Parameters:
    ///   - slot:     AIM slot number
    ///   - endpoint: Endpoint name, e.g. "infer"
    ///   - body:     Request body dict — will be JSON serialized
    public func execute(
        slot:     Int,
        endpoint: String,
        body:     [String: Any]
    ) async -> HyperCycleResult<[String: Any]> {
        return await post(path: "/aim/\(slot)/\(endpoint)", body: body)
    }

    // -------------------------------------------------------------------------
    // Convenience: ping
    // -------------------------------------------------------------------------

    /// Quick connectivity check. Returns true if the node is reachable.
    public func ping() async -> Bool {
        let result = await get(path: "/info")
        if case .success = result { return true }
        return false
    }

    // -------------------------------------------------------------------------
    // Internal helpers
    // -------------------------------------------------------------------------

    private func parseAIM(_ a: [String: Any]) -> AIMInfo {
        var costs: [EndpointCost] = []
        if let uriCost = a["uri_cost"] as? [String: Any] {
            for (_, endpointCosts) in uriCost {
                if let currencies = endpointCosts as? [String: Any] {
                    for (currency, costData) in currencies {
                        if let cd = costData as? [String: Any] {
                            costs.append(EndpointCost(
                                currency:      cd["currency"] as? String ?? currency,
                                fixed:         cd["fixed"] as? Double,
                                estimatedCost: cd["estimated_cost"] as? Double,
                                min:           cd["min"] as? Double,
                                max:           cd["max"] as? Double
                            ))
                        }
                    }
                }
            }
        }
        return AIMInfo(
            slot:        a["slot"]         as? Int    ?? 0,
            port:        a["port"]         as? Int    ?? 0,
            imageName:   a["image_name"]   as? String ?? "",
            imageTag:    a["image_tag"]    as? String ?? "latest",
            status:      a["status"]       as? String ?? "unknown",
            labels:      a["labels"]       as? [String: String] ?? [:],
            containerID: a["container_id"] as? String ?? "",
            costs:       costs
        )
    }

    private func get(path: String) async -> HyperCycleResult<[String: Any]> {
        guard let url = URL(string: "\(nodeURL)\(path)") else {
            return .failure(HyperCycleError("Invalid URL: \(nodeURL)\(path)"))
        }
        do {
            let (data, response) = try await session.data(from: url)
            let statusCode = (response as? HTTPURLResponse)?.statusCode
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return .failure(HyperCycleError("Invalid JSON response", statusCode: statusCode))
            }
            if let code = statusCode, code >= 400 {
                return .failure(HyperCycleError("HTTP \(code) from \(url)", statusCode: code))
            }
            return .success(json)
        } catch {
            return .failure(HyperCycleError("Request failed: \(error.localizedDescription)"))
        }
    }

    private func post(
        path:         String,
        body:         [String: Any],
        extraHeaders: [String: String] = [:]
    ) async -> HyperCycleResult<[String: Any]> {
        guard let url = URL(string: "\(nodeURL)\(path)") else {
            return .failure(HyperCycleError("Invalid URL: \(nodeURL)\(path)"))
        }
        do {
            var request            = URLRequest(url: url)
            request.httpMethod     = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            extraHeaders.forEach { request.setValue($1, forHTTPHeaderField: $0) }
            request.httpBody       = try JSONSerialization.data(withJSONObject: body)
            let (data, response)   = try await session.data(for: request)
            let statusCode         = (response as? HTTPURLResponse)?.statusCode
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return .failure(HyperCycleError("Invalid JSON response", statusCode: statusCode))
            }
            if let code = statusCode, code >= 400 {
                return .failure(HyperCycleError(
                    "HTTP \(code) from \(url): \(json)", statusCode: code
                ))
            }
            return .success(json)
        } catch {
            return .failure(HyperCycleError("Request failed: \(error.localizedDescription)"))
        }
    }
}
